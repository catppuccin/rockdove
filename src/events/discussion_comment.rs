use octocrab::models::webhook_events::{
    payload::{DiscussionCommentWebhookEventAction, DiscussionCommentWebhookEventPayload},
    WebhookEvent,
};

use crate::{colors::DISCUSSION_COLOR, embed_builder::EmbedBuilder};

// TODO: Create a PR to upstream (octocrab) to add typed events so that we don't
// need to use `.get()`, `.as_str()`, etc.

pub fn make_discussion_comment_embed(
    event: WebhookEvent,
    specifics: &DiscussionCommentWebhookEventPayload,
) -> Option<EmbedBuilder> {
    if !matches!(
        specifics.action,
        DiscussionCommentWebhookEventAction::Created
    ) {
        return None;
    }

    let repo = event
        .repository
        .expect("discussion comment events should always have a repository");

    let mut embed = EmbedBuilder::default();

    let repo_name = repo.full_name.unwrap_or(repo.name);

    embed.title(&format!(
        "[{}] New comment on discussion #{}: {}",
        repo_name,
        specifics
            .discussion
            .get("number")
            .expect("discussion comment should always have a number"),
        specifics
            .discussion
            .get("title")
            .expect("discussion should always have a title")
            .as_str()
            .expect("discussion title should always be a string"),
    ));

    embed.url(
        specifics
            .comment
            .get("html_url")
            .expect("discussion should always have an html url")
            .as_str()
            .expect("discussion html url should always be a string"),
    );

    embed.description(
        specifics
            .comment
            .get("body")
            .expect("discussion should always have a body")
            .as_str()
            .expect("discussion body should always be a string"),
    );

    embed.color(DISCUSSION_COLOR);

    Some(embed)
}

#[cfg(test)]
mod tests {
    use crate::{
        make_embed,
        tests::{embed_context, TestConfig},
    };

    #[test]
    fn created() {
        let payload = include_str!("../../fixtures/discussion_comment/created.json");
        let TestConfig {
            webhook_event: event,
            mut settings,
        } = TestConfig::new("discussion_comment", payload);

        let embed = make_embed(event)
            .expect("make_embed should succeed")
            .expect("event fixture can be turned into an embed");

        settings.set_info(&embed_context(&embed));
        settings.bind(|| insta::assert_yaml_snapshot!(embed));
    }
}
