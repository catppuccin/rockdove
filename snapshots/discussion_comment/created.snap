---
source: src/events/discussion_comment.rs
assertion_line: 88
info:
  author_name_length: 8
  colour_hex: "#FAB387"
  description_length: 640
  title_length: 68
---
embeds:
  - author:
      icon_url: "https://avatars.githubusercontent.com/u/58985301?v=4"
      name: sgoudham
      url: "https://github.com/sgoudham"
    color: 16429959
    description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut viverra laoreet sem quis sodales. Maecenas a dui at odio dapibus faucibus. Vivamus et mi elementum, dapibus diam in, eleifend nisl. Phasellus et congue nibh, sit amet fringilla est. Proin varius congue quam eu mattis. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Suspendisse potenti. Morbi scelerisque sapien justo, eu varius augue interdum et. Nunc elementum leo orci, quis fermentum urna semper nec.\r\n\r\nAenean porttitor semper lorem vel porttitor. Donec a sem velit. Mauris imperdiet enim nec convallis placerat. Aenean non urna..."
    title: "[catppuccin-rfc/polybar] New comment on discussion #19: another test"
    url: "https://github.com/catppuccin-rfc/polybar/discussions/19#discussioncomment-10715667"
